File: python/ktx-daemon/src/ktx_daemon/telemetry/exception.py
from __future__ import annotations

import json
import os
import re
import sys
from collections.abc import Mapping, Sequence
from pathlib import Path
from typing import Any

from ktx_daemon import VERSION
from ktx_daemon.telemetry.emitter import POSTHOG_HOST, POSTHOG_PROJECT_API_KEY
from ktx_daemon.telemetry.events import _common_envelope
from ktx_daemon.telemetry.identity import load_telemetry_identity

_KTX_REPORTED_ATTR = "__ktx_posthog_exception_reported"


def _debug_enabled(env: Mapping[str, str]) -> bool:
    return env.get("KTX_TELEMETRY_DEBUG") == "1"


def _host(env: Mapping[str, str]) -> str:
    return env.get("KTX_TELEMETRY_ENDPOINT") or POSTHOG_HOST


def _redact_static(value: str) -> str:
    patterns = [
        (
            r"([a-z][a-z0-9+.-]*://[^:\s/@]+:)([^@\s/]+)(@)",
            r"\1[redacted]\3",
        ),
        (r"\b(password|pwd)=([^;&\s]+)", r"\1=[redacted]"),
        (r"\bAuthorization\s*:\s*[^\r\n,;]+", "Authorization: [redacted]"),
        (r"\bBearer\s+[A-Za-z0-9._~+/=-]+", "Bearer [redacted]"),
        (r"\b(api[_-]?key)\s*[:=]\s*([^\s,;]+)", r"\1=[redacted]"),
        (
            r"\b(KTX_[A-Z0-9_]*|[A-Z0-9_]*(?:TOKEN|SECRET))\s*[:=]\s*([^\s,;]+)",
            r"\1=[redacted]",
        ),
        (r"([?&](?:X-Amz-Signature|X-Goog-Signature|sig)=)[^&\s]+", r"\1[redacted]"),
    ]
    redacted = value
    for pattern, replacement in patterns:
        redacted = re.sub(pattern, replacement, redacted, flags=re.IGNORECASE)
    return redacted


def _redact_text(value: str, secrets: Sequence[str]) -> str:
    redacted = value
    for secret in secrets:
        if secret:
            redacted = redacted.replace(secret, "[redacted]")
    return _redact_static(redacted)


def _clone_exception(exception: BaseException, secrets: Sequence[str]) -> BaseException:
    redacted_args = [_redact_text(str(arg), secrets) for arg in exception.args]
    try:
        cloned = type(exception)(*redacted_args)
    except Exception:
        cloned = RuntimeError(_redact_text(str(exception), secrets))
    cloned.__traceback__ = exception.__traceback__
    cloned.__cause__ = (
        _clone_exception(exception.__cause__, secrets) if exception.__cause__ else None
    )
    cloned.__context__ = (
        _clone_exception(exception.__context__, secrets)
        if exception.__context__
        else None
    )
    return cloned


def _should_skip_as_reported(exception: BaseException) -> bool:
    if getattr(exception, _KTX_REPORTED_ATTR, False):
        return True
    try:
        setattr(exception, _KTX_REPORTED_ATTR, True)
    except Exception:
        return False
    return False


def _properties(*, source: str, handled: bool, fatal: bool) -> dict[str, Any]:
    return {
        **_common_envelope(),
        "daemonVersion": os.environ.get("KTX_DAEMON_VERSION", VERSION),
        "source": source,
        "handled": handled,
        "fatal": fatal,
    }


def report_exception(
    exception: BaseException,
    *,
    source: str,
    handled: bool,
    fatal: bool,
    project_id: str | None = None,
    home_dir: Path | None = None,
    env: Mapping[str, str] | None = None,
    redaction_secrets: Sequence[str] | None = None,
) -> None:
    source_env = env if env is not None else os.environ
    try:
        identity = load_telemetry_identity(home_dir=home_dir, env=source_env)
        if not identity.enabled or not identity.install_id:
            return

        if _should_skip_as_reported(exception):
            return

        properties = _properties(source=source, handled=handled, fatal=fatal)
        groups = {"project": project_id} if project_id else None
        safe_exception = _clone_exception(exception, redaction_secrets or [])

        if _debug_enabled(source_env):
            sys.stderr.write(
                "[telemetry-exception] "
                + json.dumps(
                    {
                        "distinctId": identity.install_id,
                        "message": str(safe_exception),
                        "properties": properties,
                        "groups": groups,
                    },
                    sort_keys=True,
                )
                + "\n"
            )
            return

        if not POSTHOG_PROJECT_API_KEY.strip() or not _host(source_env).strip():
            return

        from posthog import Posthog

        client = Posthog(
            POSTHOG_PROJECT_API_KEY,
            host=_host(source_env),
            flush_at=1,
            flush_interval=0,
            sync_mode=True,
            timeout=1,
        )
        client.capture_exception(
            safe_exception,
            distinct_id=identity.install_id,
            properties=properties,
            groups=groups,
        )
        client.shutdown()
    except Exception:
        return

File: packages/cli/src/telemetry/exception.ts
import { inspect } from 'node:util';

import { getKtxCliPackageInfo, type KtxCliIo, type KtxCliPackageInfo } from '../cli-runtime.js';
import { buildCommonEnvelope } from './events.js';
import { trackTelemetryException } from './emitter.js';
import { computeTelemetryProjectId, loadTelemetryIdentity } from './identity.js';

export interface ExceptionContext {
  source: string;
  handled: boolean;
  fatal: boolean;
  extra?: Record<string, string | number | boolean>;
}

type AnyObject = object;

const reportedObjects = new WeakSet<AnyObject>();
const recentHandledPrimitives: string[] = [];
const RECENT_PRIMITIVE_LIMIT = 128;

function primitiveKey(value: unknown): string {
  return `${typeof value}:${String(value)}`;
}

function rememberHandledPrimitive(value: unknown): void {
  recentHandledPrimitives.push(primitiveKey(value));
  if (recentHandledPrimitives.length > RECENT_PRIMITIVE_LIMIT) {
    recentHandledPrimitives.splice(0, recentHandledPrimitives.length - RECENT_PRIMITIVE_LIMIT);
  }
}

function consumeHandledPrimitive(value: unknown): boolean {
  const key = primitiveKey(value);
  const index = recentHandledPrimitives.indexOf(key);
  if (index < 0) {
    return false;
  }
  recentHandledPrimitives.splice(index, 1);
  return true;
}

function shouldSkipAsAlreadyReported(error: unknown, handled: boolean): boolean {
  if ((typeof error === 'object' || typeof error === 'function') && error !== null) {
    if (reportedObjects.has(error)) {
      return true;
    }
    reportedObjects.add(error);
    return false;
  }

  if (handled) {
    rememberHandledPrimitive(error);
    return false;
  }

  return consumeHandledPrimitive(error);
}

function escapeRegExp(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function redactStaticPatterns(value: string): string {
  return value
    .replace(/([a-z][a-z0-9+.-]*:\/\/[^:\s/@]+:)([^@\s/]+)(@)/gi, '$1[redacted]$3')
    .replace(/\b(password|pwd)=([^;&\s]+)/gi, '$1=[redacted]')
    .replace(/\bAuthorization\s*:\s*[^\r\n,;]+/gi, 'Authorization: [redacted]')
    .replace(/\bBearer\s+[A-Za-z0-9._~+/=-]+/gi, 'Bearer [redacted]')
    .replace(/\b(api[_-]?key)\s*[:=]\s*([^\s,;]+)/gi, '$1=[redacted]')
    .replace(/\b(KTX_[A-Z0-9_]*|[A-Z0-9_]*(?:TOKEN|SECRET))\s*[:=]\s*([^\s,;]+)/g, '$1=[redacted]')
    .replace(/([?&](?:X-Amz-Signature|X-Goog-Signature|sig)=)[^&\s]+/gi, '$1[redacted]');
}

function redactText(value: string, secrets: ReadonlyArray<string>): string {
  let redacted = value;
  for (const secret of secrets) {
    if (secret) {
      redacted = redacted.replace(new RegExp(escapeRegExp(secret), 'g'), '[redacted]');
    }
  }
  return redactStaticPatterns(redacted);
}

const FORBIDDEN_EXTRA_PROPERTY_KEYS = new Set([
  'argv',
  'args',
  'env',
  'environment',
  'sql',
  'query',
  'prompt',
  'mcparguments',
  'mcpargs',
  'tablename',
  'schemaname',
  'columnname',
  'databaseurl',
  'connectionstring',
  'url',
  'password',
  'token',
  'apikey',
  'api_key',
  'authorization',
]);

function safeExtraProperties(
  extra: Record<string, string | number | boolean> | undefined,
): Record<string, string | number | boolean> {
  const safe: Record<string, string | number | boolean> = {};
  for (const [key, value] of Object.entries(extra ?? {})) {
    if (!FORBIDDEN_EXTRA_PROPERTY_KEYS.has(key.replace(/[^a-z0-9_]/gi, '').toLowerCase())) {
      safe[key] = value;
    }
  }
  return safe;
}

function toMessage(error: unknown): string {
  if (error instanceof Error) {
    return error.message;
  }
  if (typeof error === 'string') {
    return error;
  }
  return inspect(error, { depth: 4, breakLength: 120 });
}

function sanitizedError(error: unknown, secrets: ReadonlyArray<string>): Error {
  if (error instanceof Error) {
    const cause = 'cause' in error ? (error as Error & { cause?: unknown }).cause : undefined;
    const clone = new Error(redactText(error.message, secrets), {
      ...(cause !== undefined ? { cause: sanitizedError(cause, secrets) } : {}),
    });
    clone.name = error.name;
    if (error.stack) {
      clone.stack = redactText(error.stack, secrets);
    }
    return clone;
  }
  return new Error(redactText(toMessage(error), secrets));
}

export async function reportException(input: {
  error: unknown;
  context: ExceptionContext;
  io: KtxCliIo;
  packageInfo?: KtxCliPackageInfo;
  projectDir?: string;
  immediate?: boolean;
  redactionSecrets?: ReadonlyArray<string>;
}): Promise<void> {
  try {
    if (shouldSkipAsAlreadyReported(input.error, input.context.handled)) {
      return;
    }

    const debug = process.env.KTX_TELEMETRY_DEBUG === '1';
    const identity = await loadTelemetryIdentity({
      stderr: input.io.stderr,
      env: process.env,
    });

    if ((!identity.enabled || !identity.installId) && !debug) {
      return;
    }

    const packageInfo = input.packageInfo ?? getKtxCliPackageInfo();
    const installId = identity.installId ?? 'debug';
    const projectId = input.projectDir ? computeTelemetryProjectId(installId, input.projectDir) : undefined;
    const safeError = sanitizedError(input.error, input.redactionSecrets ?? []);
    const properties: Record<string, unknown> = {
      ...buildCommonEnvelope({
        cliVersion: packageInfo.version,
        isCi: Boolean(process.env.CI),
      }),
      source: input.context.source,
      handled: input.context.handled,
      fatal: input.context.fatal,
      ...(projectId ? { projectId } : {}),
      ...safeExtraProperties(input.context.extra),
    };

    delete properties.$groups;
    await trackTelemetryException({
      error: safeError,
      distinctId: installId,
      properties,
      env: process.env,
      stderr: input.io.stderr,
      immediate: input.immediate,
    });
  } catch {
    return;
  }
}

/** @internal */
export function __resetTelemetryExceptionStateForTests(): void {
  recentHandledPrimitives.length = 0;
}
